Add `-Z teach` flag to provide extended diagnostic help

estebank opened the pull request:
First step to have a way for the compiler to provide as much information as the `--explain E0XXX` flag does, but providing the text in the context of the users' code.

r? @nikomatsakis

estebank committed:
Add `--explain` for extended error explanations

estebank committed:
Mark `--explain` as unstable

estebank committed:
Add explanation for E0607

estebank committed:
Fix test for variadic error change

estebank committed:
Only emit expanded diagnostic information once

estebank commented:
<img width="880" alt="screen shot 2018-01-21 at 21 39 05" src="https://user-images.githubusercontent.com/1606434/35206728-90d9300c-fef3-11e7-9a45-2a33ea0b048f.png">

estebank committed:
Fix test redux

nikomatsakis commented:
Maybe we need a different name (for back-compat).

Here are some synonyms:

`--expound`, `--say-more`, `--details`, `--teach`, `--elucidate`

I think I like `--teach`.

nikomatsakis commented:
Also, long term, do you envision us having this text etc in the code like this, or separated into some form of "template"?

estebank commented:
At the very least I would separate the place where the error is being emitted from the error being prepared for presentation. Using a template is a good possibility, but I feel it can be approached independently from the introduction of the flag.

The way to use it now is `rustc -Z unstable-options -Z explain file.rs`. I feel that using the same flag shouldn't as now shouldn't be a problem, as in both cases the intention is the same, give me more context on what I'm seeing. Also, the content should be, of not the same, equivalent for both, to the point where `explain E0XXX` should look like diagnostic output too.

nikomatsakis commented:
Do we have a tracking issue for this feature? If so, we ought to start documenting these questions.

zackmdavis commented:
> Using a template is a good possibility, but I feel it can be approached independently from the introduction of the flag.

We should keep in mind that deferring work to the future is only a good idea when the minimal PR leaves the code base in a _better_ state than it was previously (even if it's not optimal); we don't want giant explanations cluttering up our code forever if no one gets around to implementing templates in a timely manner.

estebank committed:
Don't make it necessary to enable `unstable-options`

estebank committed:
Create `StructuredDiagnostic`

Create the concept of an `StructuredDiagnostic` that is self-contained
with enough knowledge of all variables to create a `DiagnosticBuilder`,
including different possible versions (one line output and expanded
explanations).

estebank commented:
https://github.com/rust-lang/rust/pull/47652/commits/19d16379793a98e4e0b9d8f63b89dc95a6118273 is an early mockup of what `StructuredDiagnostic`s would look like. Right now I'm not too happy with how it looks, as it is slightly more verbose than the previous code, but I think it works as a starting point for talking and iterating upon. Having all the diagnostic errors that could use templates in one place would make it easier to migrate them once such a solution is written.

nikomatsakis commented:
The more I think about it, the more I like `--teach`. I think we should call this `-Zteach` for now =)

nikomatsakis commented on src/librustc_typeck/check/cast.rs:
```diff
@@ -281,35 +281,12 @@ impl<'a, 'gcx, 'tcx> CastCheck<'tcx> {
-                }
-                err.emit();
+                use structured_errors::{SizedUnsizedCastError, StructuredDiagnostic};
+                SizedUnsizedCastError::new(&fcx.tcx.sess,
```
Personally, I'd rather something with labels I think, like:

```
SizedUnsizedCastError {
    span: self.span,
    thin_ty: self.expr_ty,
    fat_ty: fcx.ty_to_string(self.cast_ty) // although if we could do `self.cast_ty` it'd be even better
}.emit(fcx.tcx);
```

These labels would then also appear in the templating logic.

estebank replied:
I'm thinking that the templating could look something like the following:

```
diagnostic_template!(
Any independent text will be notes.

Such as this one and the previous one.

Here is where a span would appear:
$secondary_span the label for the span immediately after

I haven't thought about how to deal with suggestions.
)
```
```
error[E0XXX]: error message outside of template
  --> $FILE:21:9
   |
21 |         foo();
   |         ^^^^^
   = note: Any independent text will be notes.

          Such as this one and the previous one.
note: Here is where a span would appear:
  --> $FILE:21:9
   |
21 |         foo();
   |         ^^^^^ the label for the span immediately after
   = note: I haven't thought about how to deal with suggestions.
```

In the background it would convert the template string into calls to `err.*()`.

How does that sound?

nikomatsakis replied:
Interesting. Quite different than what I had in mind, but I like it as a start, and maybe more practical than some of my more grandiose ideas =)

I think one key question is whether the `--teach` output would still have the "basic structure" of the messages that we've been giving (as in this PR), or whether it would change to something more essay like. 

I think I had initially imagined that, in `--teach` mode, we would change "our look" somewhat drastically. e.g (using E0010 as an example):

```
error[E0010]: boxed values cannot be used in statics
  --> $FILE:21:9
  |
21 |         const CON : Box<i32> = box 0;
   |                                ^^^

The value of statics and constants must be known at compile time, and they live
for the entire lifetime of a program. Creating a boxed value allocates memory on
the heap at runtime, and therefore cannot be done at compile time.
```

For more complex examples, there might be multiple snippets:

```
error[E0010]: shadowed constant `CON`
  --> $FILE:21:9
  |
21 |         const CON : Box<i32> = box 0;
   |               ^^^

This code is illegal. The problem is that the const CON is shadowed by another constant declared earlier, also named `CON`:

  --> $FILE:18:9
  |
18 |         const CON : Box<i32> = box 22;
   |               ^^^
```

estebank replied:
The only reason I wasn't considering moving in that direction is because I though we wouldn't want to present the same explanation over and over again, so mixing the current diagnostics with the teach diagnostics would be a bit jarring. But if we don't mind either mixing both styles or repeating walls of text, then [I don't see why we wouldn't do it](https://github.com/rust-lang/rfcs/blob/master/text/1644-default-and-expanded-rustc-errors.md#expanded-error-messages).

estebank replied:
Also, I think that the differences in output could act like a stylesheet, where the `DiagnosticBuilder` is still created in the same way, but when emitting it to the output we change how things are presented depending on wether we're in `teach` mode or not, by removing the left margin on `note`s, line wrapping, how we deal with suggestions, maybe [even highlighting code](https://github.com/rust-lang/rust/pull/45752) based on this.

estebank committed:
Rename `-Z explain` to `-Z teach`

nikomatsakis commented on src/librustc_errors/lib.rs:
```diff
@@ -244,6 +244,7 @@ pub struct Handler {
     continue_after_error: Cell<bool>,
     delayed_span_bug: RefCell<Option<Diagnostic>>,
     tracked_diagnostics: RefCell<Option<Vec<Diagnostic>>>,
+    tracked_diagnostic_codes: RefCell<FxHashSet<DiagnosticId>>,
```
can we put a comment on this field explaining what it's for?

nikomatsakis commented on src/librustc_errors/lib.rs:
```diff
@@ -575,13 +577,21 @@ impl Handler {
         (ret, diagnostics)
     }
 
+    pub fn code_emitted(&self, code: &DiagnosticId) -> bool {
```
comment (short one would suffice)

e.g., `/// True if a diagnostic with this code has already been emitted. Used to suppress duplicate messages with -Zteach.`

nikomatsakis reviewed (changes requested):
OK, good start. I just left a nit that I think a comment would be good (not that this is a *particularly* confusing case, just that I think we should start to comment more)

r=me with that

nikomatsakis commented:
I think however we should think about some kind of public document laying out the desired design here at some point. We could probably get a bunch of people involved in making this happen if we do it right!

estebank committed:
Add description to field and method

estebank commented:
@bors r=nikomatsakis

bors commented:
:pushpin: Commit 2b73733 has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis 2b737334961916daee73ea018eea877f389ad0dc -->
